atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -389,36 +462,55 @@ def do_worker_build(self, release, cluster_info, koji_upload_dir, task_id):
                 except OsbsException:
                     pass
 
+    def select_and_start_cluster(self, platform):
+        ''' Choose a cluster and start a build on it '''
+
+        config = get_config(self.workflow)
+        clusters = config.get_enabled_clusters_for_platform(platform)
+
+        if not clusters:
+            raise UnknownPlatformException('No clusters found for platform {}!'
+                                           .format(platform))
+
+        retry_contexts = {
+            cluster.name: ClusterRetryContext(self.max_cluster_fails)
+            for cluster in clusters
+        }
+
+        while True:
+            possible_cluster_info = self.get_clusters(platform, retry_contexts, clusters)
+
+            for cluster_info in possible_cluster_info:
+                ctx = retry_contexts[cluster_info.cluster.name]
+                try:
+                    self.do_worker_build(cluster_info)
+                    return
```

Comment:
Looking at `do_worker_build()`, I don't think it will ever raise `OsbsException`. If creating the build fails due to `OsbsResponseException` it looks like it will just create a `WorkerBuildInfo` with build=None.

Doesn't the `return` here assume that `do_worker_build()` succeeded in creating a build?

Do we have a unit test to verify this situation is handled correctly?

Reply:
Ha! Good catch. Let's change `do_worker_build` to not catch exceptions when `cluster_info.osbs.create_worker_build(**kwargs)` is called. This should cause the exception to propagate as needed.
What do you think?

Reply:
Should do_worker_build no longer try and catch? Or should it just reraise OsbsException and handle all others in its usual way?

Reply:
I think we want to have do_worker_build()
* always append to self.worker_builds, as it does now
* raise any exception that occurs when starting a build or monitoring it

So where it currently catches `Exception`, I think it should note what the exception is and arrange for `build_info.monitor_exception` to be set to it. This may mean having WorkerBuildInfo's constructor not take 'build' but instead just set it to None, and have do_worker_build() setting build_info.build when it's known.

Then, when monitoring the build, it should catch the exception as before and set build_info.monitor_exception to it. It should also try to cancel the build, as it currently does. But at the end of all that, it should raise that exception.

@lcarva, what do you think?

Reply:
We discussed this and came to the conclusion that catching and re-raising OsbsException is the better approach.